File: src/constfold.ts
// Compile-time constant folding for primitive constant expressions (JLS 15.28),
// matching javac: an expression built from literals and operators is evaluated at
// compile time so the emitter can push the folded constant. Integer arithmetic
// wraps in two's complement (32-bit int, 64-bit long). Float/double/char/String
// and `final` constant variables are not folded yet.

import {
  type BinaryExpression,
  type LiteralExpression,
  type Node,
  type PrefixUnaryExpression,
  SyntaxKind,
} from "./types.ts";

export type ConstValue =
  | { readonly kind: "int"; readonly value: bigint }
  | { readonly kind: "long"; readonly value: bigint }
  | { readonly kind: "boolean"; readonly value: boolean };

const int32 = (v: bigint): bigint => BigInt.asIntN(32, v);
const int64 = (v: bigint): bigint => BigInt.asIntN(64, v);

function parseIntLiteral(text: string): bigint | undefined {
  const t = text.replace(/_/g, "");
  try {
    if (/^0[0-7]+$/.test(t)) return BigInt(parseInt(t, 8)); // legacy octal
    return BigInt(t); // handles decimal, 0x..., 0b...
  } catch {
    return undefined;
  }
}

function num(value: ConstValue | undefined): { kind: "int" | "long"; value: bigint } | undefined {
  return value && value.kind !== "boolean" ? value : undefined;
}

function wrap(kind: "int" | "long", value: bigint): bigint {
  return kind === "long" ? int64(value) : int32(value);
}

function foldPrefix(node: PrefixUnaryExpression): ConstValue | undefined {
  const operand = foldConstant(node.operand);
  if (!operand) return undefined;
  switch (node.operator) {
    case SyntaxKind.PlusToken:
      return num(operand);
    case SyntaxKind.MinusToken: {
      const n = num(operand);
      return n && { kind: n.kind, value: wrap(n.kind, -n.value) };
    }
    case SyntaxKind.TildeToken: {
      const n = num(operand);
      return n && { kind: n.kind, value: wrap(n.kind, -n.value - 1n) };
    }
    case SyntaxKind.ExclamationToken:
      return operand.kind === "boolean" ? { kind: "boolean", value: !operand.value } : undefined;
    default:
      return undefined;
  }
}

const COMPARE = {
  [SyntaxKind.LessThanToken]: (a, b) => a < b,
  [SyntaxKind.LessThanEqualsToken]: (a, b) => a <= b,
  [SyntaxKind.GreaterThanToken]: (a, b) => a > b,
  [SyntaxKind.GreaterThanEqualsToken]: (a, b) => a >= b,
  [SyntaxKind.EqualsEqualsToken]: (a, b) => a === b,
  [SyntaxKind.ExclamationEqualsToken]: (a, b) => a !== b,
} as const satisfies Partial<Record<SyntaxKind, (a: bigint, b: bigint) => boolean>>;

function foldBinary(node: BinaryExpression): ConstValue | undefined {
  const left = foldConstant(node.left);
  const right = foldConstant(node.right);
  if (!left || !right) return undefined;
  const op = node.operatorToken;

  if (left.kind === "boolean" && right.kind === "boolean") {
    switch (op) {
      case SyntaxKind.AmpersandAmpersandToken:
      case SyntaxKind.AmpersandToken:
        return { kind: "boolean", value: left.value && right.value };
      case SyntaxKind.BarBarToken:
      case SyntaxKind.BarToken:
        return { kind: "boolean", value: left.value || right.value };
      case SyntaxKind.CaretToken:
        return { kind: "boolean", value: left.value !== right.value };
      case SyntaxKind.EqualsEqualsToken:
        return { kind: "boolean", value: left.value === right.value };
      case SyntaxKind.ExclamationEqualsToken:
        return { kind: "boolean", value: left.value !== right.value };
      default:
        return undefined;
    }
  }

  const a = num(left);
  const b = num(right);
  if (!a || !b) return undefined;

  const compare = COMPARE[op as keyof typeof COMPARE];
  if (compare) return { kind: "boolean", value: compare(a.value, b.value) };

  const kind = a.kind === "long" || b.kind === "long" ? "long" : "int";
  switch (op) {
    case SyntaxKind.PlusToken:
      return { kind, value: wrap(kind, a.value + b.value) };
    case SyntaxKind.MinusToken:
      return { kind, value: wrap(kind, a.value - b.value) };
    case SyntaxKind.AsteriskToken:
      return { kind, value: wrap(kind, a.value * b.value) };
    case SyntaxKind.SlashToken:
      return b.value === 0n ? undefined : { kind, value: wrap(kind, a.value / b.value) };
    case SyntaxKind.PercentToken:
      return b.value === 0n ? undefined : { kind, value: wrap(kind, a.value % b.value) };
    case SyntaxKind.AmpersandToken:
      return { kind, value: wrap(kind, a.value & b.value) };
    case SyntaxKind.BarToken:
      return { kind, value: wrap(kind, a.value | b.value) };
    case SyntaxKind.CaretToken:
      return { kind, value: wrap(kind, a.value ^ b.value) };
    case SyntaxKind.LessThanLessThanToken:
    case SyntaxKind.GreaterThanGreaterThanToken:
    case SyntaxKind.GreaterThanGreaterThanGreaterThanToken: {
      // A shift's result type and the distance mask come from the (promoted) left
      // operand only; the right operand never widens the result (JLS 15.19).
      const sk = a.kind;
      const sb = sk === "long" ? 64n : 32n;
      const dist = b.value & (sb - 1n);
      if (op === SyntaxKind.LessThanLessThanToken) {
        return { kind: sk, value: wrap(sk, a.value << dist) };
      }
      if (op === SyntaxKind.GreaterThanGreaterThanToken) {
        return { kind: sk, value: wrap(sk, a.value >> dist) };
      }
      return { kind: sk, value: wrap(sk, BigInt.asUintN(Number(sb), a.value) >> dist) };
    }
    default:
      return undefined;
  }
}

/** Evaluate a primitive constant expression, or undefined if it is not constant. */
export function foldConstant(node: Node): ConstValue | undefined {
  switch (node.kind) {
    case SyntaxKind.ParenthesizedExpression:
      return foldConstant((node as unknown as { expression: Node }).expression);
    case SyntaxKind.NumericLiteral: {
      const text = (node as LiteralExpression).value;
      const t = text.replace(/_/g, "");
      const isHexOrBin = /^0[xXbB]/.test(t);
      // Decimal float/double, or hex floating-point: not an integer constant.
      // (In hex/binary literals a-f are digits, so only guard them on a 'p' exponent.)
      if (isHexOrBin ? /[pP]/.test(t) : /[.eEfFdD]/.test(t)) return undefined;
      const isLong = /[lL]$/.test(text);
      const parsed = parseIntLiteral(isLong ? text.replace(/[lL]$/, "") : text);
      if (parsed === undefined) return undefined;
      return isLong
        ? { kind: "long", value: int64(parsed) }
        : { kind: "int", value: int32(parsed) };
    }
    case SyntaxKind.TrueKeyword:
      return { kind: "boolean", value: true };
    case SyntaxKind.FalseKeyword:
      return { kind: "boolean", value: false };
    case SyntaxKind.PrefixUnaryExpression:
      return foldPrefix(node as PrefixUnaryExpression);
    case SyntaxKind.BinaryExpression:
      return foldBinary(node as BinaryExpression);
    default:
      return undefined;
  }
}

File: togo/internal/compiler/constfold.go
package compiler

// Compile-time constant folding for primitive constant expressions (JLS 15.28),
// matching javac: an expression built from literals and operators is evaluated at
// compile time so the emitter can push the folded constant. Integer arithmetic
// wraps in two's complement (32-bit int, 64-bit long). Float/double/char/String
// and `final` constant variables are not folded yet. Port of src/compiler/constfold.ts.

import (
	"strconv"
	"strings"
)

type ConstKind int

const (
	ConstInt ConstKind = iota
	ConstLong
	ConstBool
)

// ConstValue is a folded primitive constant. For int/long the value lives in Int;
// for boolean in Bool. A nil *ConstValue means "not a constant expression".
type ConstValue struct {
	Kind ConstKind
	Int  int64
	Bool bool
}

func wrap(kind ConstKind, v int64) int64 {
	if kind == ConstLong {
		return v
	}
	return int64(int32(v))
}

// parseIntLiteral parses a Java integer literal magnitude into its 64-bit pattern.
func parseIntLiteral(text string) (uint64, bool) {
	t := strings.ReplaceAll(text, "_", "")
	// Legacy octal: 0 followed by octal digits.
	if len(t) >= 2 && t[0] == '0' && isOctalDigits(t[1:]) {
		v, err := strconv.ParseUint(t[1:], 8, 64)
		return v, err == nil
	}
	if len(t) >= 2 && (t[1] == 'x' || t[1] == 'X') && t[0] == '0' {
		v, err := strconv.ParseUint(t[2:], 16, 64)
		return v, err == nil
	}
	if len(t) >= 2 && (t[1] == 'b' || t[1] == 'B') && t[0] == '0' {
		v, err := strconv.ParseUint(t[2:], 2, 64)
		return v, err == nil
	}
	v, err := strconv.ParseUint(t, 10, 64)
	return v, err == nil
}

func isOctalDigits(s string) bool {
	for i := 0; i < len(s); i++ {
		if s[i] < '0' || s[i] > '7' {
			return false
		}
	}
	return len(s) > 0
}

func numKind(v *ConstValue) (ConstValue, bool) {
	if v != nil && v.Kind != ConstBool {
		return *v, true
	}
	return ConstValue{}, false
}

func foldPrefix(node *Node) *ConstValue {
	d := node.AsPrefixUnaryExpression()
	operand := FoldConstant(d.Operand)
	if operand == nil {
		return nil
	}
	switch d.Operator {
	case PlusToken:
		if n, ok := numKind(operand); ok {
			return &n
		}
		return nil
	case MinusToken:
		if n, ok := numKind(operand); ok {
			return &ConstValue{Kind: n.Kind, Int: wrap(n.Kind, -n.Int)}
		}
		return nil
	case TildeToken:
		if n, ok := numKind(operand); ok {
			return &ConstValue{Kind: n.Kind, Int: wrap(n.Kind, -n.Int-1)}
		}
		return nil
	case ExclamationToken:
		if operand.Kind == ConstBool {
			return &ConstValue{Kind: ConstBool, Bool: !operand.Bool}
		}
		return nil
	default:
		return nil
	}
}

func foldCompare(op SyntaxKind, a, b int64) (bool, bool) {
	switch op {
	case LessThanToken:
		return a < b, true
	case LessThanEqualsToken:
		return a <= b, true
	case GreaterThanToken:
		return a > b, true
	case GreaterThanEqualsToken:
		return a >= b, true
	case EqualsEqualsToken:
		return a == b, true
	case ExclamationEqualsToken:
		return a != b, true
	default:
		return false, false
	}
}

func foldBinary(node *Node) *ConstValue {
	d := node.AsBinaryExpression()
	left := FoldConstant(d.Left)
	right := FoldConstant(d.Right)
	if left == nil || right == nil {
		return nil
	}
	op := d.OperatorToken

	if left.Kind == ConstBool && right.Kind == ConstBool {
		switch op {
		case AmpersandAmpersandToken, AmpersandToken:
			return &ConstValue{Kind: ConstBool, Bool: left.Bool && right.Bool}
		case BarBarToken, BarToken:
			return &ConstValue{Kind: ConstBool, Bool: left.Bool || right.Bool}
		case CaretToken, ExclamationEqualsToken:
			return &ConstValue{Kind: ConstBool, Bool: left.Bool != right.Bool}
		case EqualsEqualsToken:
			return &ConstValue{Kind: ConstBool, Bool: left.Bool == right.Bool}
		default:
			return nil
		}
	}

	a, aok := numKind(left)
	b, bok := numKind(right)
	if !aok || !bok {
		return nil
	}

	if result, ok := foldCompare(op, a.Int, b.Int); ok {
		return &ConstValue{Kind: ConstBool, Bool: result}
	}

	kind := ConstInt
	if a.Kind == ConstLong || b.Kind == ConstLong {
		kind = ConstLong
	}
	switch op {
	case PlusToken:
		return &ConstValue{Kind: kind, Int: wrap(kind, a.Int+b.Int)}
	case MinusToken:
		return &ConstValue{Kind: kind, Int: wrap(kind, a.Int-b.Int)}
	case AsteriskToken:
		return &ConstValue{Kind: kind, Int: wrap(kind, a.Int*b.Int)}
	case SlashToken:
		if b.Int == 0 {
			return nil
		}
		return &ConstValue{Kind: kind, Int: wrap(kind, a.Int/b.Int)}
	case PercentToken:
		if b.Int == 0 {
			return nil
		}
		return &ConstValue{Kind: kind, Int: wrap(kind, a.Int%b.Int)}
	case AmpersandToken:
		return &ConstValue{Kind: kind, Int: wrap(kind, a.Int&b.Int)}
	case BarToken:
		return &ConstValue{Kind: kind, Int: wrap(kind, a.Int|b.Int)}
	case CaretToken:
		return &ConstValue{Kind: kind, Int: wrap(kind, a.Int^b.Int)}
	case LessThanLessThanToken, GreaterThanGreaterThanToken, GreaterThanGreaterThanGreaterThanToken:
		// A shift's result type and the distance mask come from the (promoted) left
		// operand only; the right operand never widens the result (JLS 15.19).
		sk := a.Kind
		var sb int64 = 32
		if sk == ConstLong {
			sb = 64
		}
		dist := uint64(b.Int & (sb - 1))
		switch op {
		case LessThanLessThanToken:
			return &ConstValue{Kind: sk, Int: wrap(sk, a.Int<<dist)}
		case GreaterThanGreaterThanToken:
			return &ConstValue{Kind: sk, Int: wrap(sk, a.Int>>dist)}
		default: // >>> logical: zero-fill from the unsigned operand
			if sk == ConstLong {
				return &ConstValue{Kind: sk, Int: wrap(sk, int64(uint64(a.Int)>>dist))}
			}
			return &ConstValue{Kind: sk, Int: wrap(sk, int64(uint32(a.Int)>>dist))}
		}
	default:
		return nil
	}
}

// FoldConstant evaluates a primitive constant expression, or returns nil if it
// is not constant.
func FoldConstant(node *Node) *ConstValue {
	switch node.Kind {
	case ParenthesizedExpression:
		return FoldConstant(node.AsParenthesizedExpression().Expression)
	case NumericLiteral:
		text := node.AsLiteralExpression().Value
		t := strings.ReplaceAll(text, "_", "")
		isHexOrBin := len(t) >= 2 && t[0] == '0' && (t[1] == 'x' || t[1] == 'X' || t[1] == 'b' || t[1] == 'B')
		// Decimal float/double, or hex floating-point: not an integer constant.
		// (In hex/binary literals a-f are digits, so only guard them on a 'p' exponent.)
		if isHexOrBin {
			if strings.ContainsAny(t, "pP") {
				return nil
			}
		} else if strings.ContainsAny(t, ".eEfFdD") {
			return nil
		}
		isLong := len(text) > 0 && (text[len(text)-1] == 'l' || text[len(text)-1] == 'L')
		lit := text
		if isLong {
			lit = text[:len(text)-1]
		}
		parsed, ok := parseIntLiteral(lit)
		if !ok {
			return nil
		}
		if isLong {
			return &ConstValue{Kind: ConstLong, Int: int64(parsed)}
		}
		return &ConstValue{Kind: ConstInt, Int: int64(int32(uint32(parsed)))}
	case TrueKeyword:
		return &ConstValue{Kind: ConstBool, Bool: true}
	case FalseKeyword:
		return &ConstValue{Kind: ConstBool, Bool: false}
	case PrefixUnaryExpression:
		return foldPrefix(node)
	case BinaryExpression:
		return foldBinary(node)
	default:
		return nil
	}
}
